iothub/device/src/PlugAndPlay/PnpHelper.cs

```diff
@@ -1,91 +1,148 @@
 using System.Text;
-using System.Text.RegularExpressions;
-using Microsoft.Azure.Devices.Client;
+using Microsoft.Azure.Devices.Client.Extensions;
 using Microsoft.Azure.Devices.Shared;
 using Newtonsoft.Json.Linq;
 
-namespace PnpHelpers
+namespace Microsoft.Azure.Devices.Client.PlugAndPlay
 {
+    /// <summary>
+    /// A helper class for formatting messages and properties as per plug and play convention.
+    /// </summary>
     public static class PnpHelper
     {
-        private const string TelemetryComponentPropertyName = "$.sub";
-        private const string EncodingUtf8 = "utf-8";
-        private const string ContentApplicationJson = "application/json";
-        private const string PropertyComponentIdentifierKey = "__t";
-        private const string PropertyComponentIdentifierValue = "c";
+        /// <summary>
+        /// The content type for a plug and play compatible telemetry message.
+        /// </summary>
+        public const string ContentApplicationJson = "application/json";
+
+        /// <summary>
+        /// The key for a component identifier within a property update patch. Corresponding value is <see cref="PropertyComponentIdentifierValue"/>.
+        /// </summary>
+        public const string PropertyComponentIdentifierKey = "__t";
+
+        /// <summary>
+        /// The value for a component identifier within a property update patch. Corresponding key is <see cref="PropertyComponentIdentifierKey"/>.
+        /// </summary>
+        public const string PropertyComponentIdentifierValue = "c";
 
         /// <summary>
         /// Create a plug and play compatible telemetry message.
         /// </summary>
         /// <param name="telemetryName">The name of the telemetry, as defined in the DTDL interface. Must be 64 characters or less. For more details refer <see href="https://github.com/Azure/opendigitaltwins-dtdl/blob/master/DTDL/v2/dtdlv2.md#telemetry"/>.</param>
         /// <param name="serializedTelemetryValue">The serialized telemetry payload, in the format defined in the DTDL interface.</param>
-        /// <param name="componentName">(optional) The name of the component in which the telemetry is defined. Can be null for telemetry defined under the root interface.</param>
+        /// <param name="componentName">The name of the component in which the telemetry is defined. Can be null for telemetry defined under the root interface.</param>
+        /// <param name="encoding">The character encoding to be used when encoding the message body to bytes. This defaults to utf-8.</param>
         /// <returns>A plug and play compatible telemetry message, which can be sent to IoT Hub.</returns>
-        public static Message CreateIothubMessageUtf8(string telemetryName, string serializedTelemetryValue, string componentName = default)
+        public static Message CreateMessage(string telemetryName, string serializedTelemetryValue, string componentName = default, Encoding encoding = default)
```

Comment:
Should encoding = default or Encoding.UTF8?

Reply:
unfortunately we can't set Encoding.UTF8 as the default value, since it isn't a compile time constant.